core/include/scipp/core/element_array.h

```diff
@@ -49,13 +49,36 @@ template <class T> class element_array {
     resize(new_size, default_init_elements);
   }
 
-  template <class InputIt,
-            std::enable_if_t<!std::is_integral<InputIt>{}, int> = 0>
-  element_array(InputIt first, InputIt last) {
+  //  template <class InputIt,
+  //            std::enable_if_t<!std::is_integral<InputIt>{}, int> = 0>
+  //  element_array(InputIt first, InputIt last) {
+  //    resize(std::distance(first, last), default_init_elements);
+  //    std::copy(first, last, data());
+  //  }
+
+  template <class Iter,
+            std::enable_if_t<
+                std::is_assignable<T &, typename Iter::value_type>{}, int> = 0>
+  element_array(Iter first, Iter last) {
     resize(std::distance(first, last), default_init_elements);
     std::copy(first, last, data());
   }
 
+  template <class U, std::enable_if_t<std::is_assignable_v<T &, U>, int> = 0>
+  element_array(U *first, U *last) {
+    resize(std::distance(first, last), default_init_elements);
+    std::copy(first, last, data());
+  }
+
+  template <class U, template <class> class Container,
+            std::enable_if_t<std::is_assignable_v<T &, U>, int> = 0>
+  element_array(Container<U> &&c) : element_array(c.begin(), c.end()) {}
```

Comment:
Not sure why this is suddenly required? Also, doesn't it bypass the built-in move assignment (if `Container == element_array` and `T == U`), thus always forcing a copy based on iterators?

Reply:
(I am not sure what the actual overload-resolution rules do in this case)

Reply:
I've put this for `Valuses(std::vector<int>(5, 6))`, for example. In principle could be removed. I'm pretty sure it wouldn't be used then U=T, there is a test, which checks that.